lkmpg.tex

```diff
@@ -681,40 +683,49 @@ \subsection{The proc\_ops Structure}
     int (*flock) (struct file *, int, struct file_lock *);
     ssize_t (*splice_write)(struct pipe_inode_info *, struct file *, loff_t *, size_t, unsigned int);
     ssize_t (*splice_read)(struct file *, loff_t *, struct pipe_inode_info *, size_t, unsigned int);
-    int (*setlease)(struct file *, long, struct file_lock **);
+    int (*setlease)(struct file *, long, struct file_lock **, void **);
     long (*fallocate)(struct file *file, int mode, loff_t offset,
-	      loff_t len);
-    int (*show_fdinfo)(struct seq_file *m, struct file *f);
-};
+        loff_t len);
+    void (*show_fdinfo)(struct seq_file *m, struct file *f);
+#ifndef CONFIG_MMU
+  unsigned (*mmap_capabilities)(struct file *);
+#endif
+  ssize_t (*copy_file_range)(struct file *, loff_t, struct file *,
+      loff_t, size_t, unsigned int);
+  loff_t (*remap_file_range)(struct file *file_in, loff_t pos_in,
+           struct file *file_out, loff_t pos_out,
+           loff_t len, unsigned int remap_flags);
+  int (*fadvise)(struct file *, loff_t, loff_t, int);
+} __randomize_layout;
 \end{code}
 
-Some operations are not implemented by a driver. For example, a driver that handles a video card won't need to read from a directory structure. The corresponding entries in the proc\_ops structure should be set to NULL.
+Some operations are not implemented by a driver. For example, a driver that handles a video card won't need to read from a directory structure. The corresponding entries in the file\_operations structure should be set to NULL.
 
 There is a gcc extension that makes assigning to this structure more convenient. You'll see it in modern drivers, and may catch you by surprise. This is what the new way of assigning to the structure looks like:
 
 \begin{code}
-struct proc_ops fops = {
```

Comment:
Check `examples/procfs1.c` and you shall explain the interface change for Linux v5.6+.

Reply:
But the `proc_fs` structure is only for /proc file system and this chapter is introducing about character device drivers, should we explain it in next chapter?

Reply:
Then, this patch should touch back and forth.